#                                     # Maji Ndogo Water Crisis Investigation

A new database of **60,000 records** captures vital details on water sources, usage, and environmental conditions. This data holds the clues needed to understand and address the nation’s growing water crisis.


## The Aim

The objective is to transform the raw dataset into **clear, actionable insights** by:  

- Exploring the data’s structure  
- Understanding its variables  
- Uncovering hidden patterns  
- Identifying urgent problems  

Each step will move from scattered numbers to solutions that inform decisions and support lasting access to clean water.




In [ ]:
%load_ext sql

#  Beginning Our Data-Driven Journey in Maji Ndogo

### 1.1 Getting to Know Our Data

**Context:** Before delving into complex analysis, it is crucial to understand the foundational tables and their structure within the Maji Ndogo database. This involves exploring the existing tables and familiarising oneself with their contents.

- [ ]  **Load the database** and pull up the first few records from each table.
- [ ]  **Identify the number and names of tables** in the database using a `SHOW TABLES` query.
- [ ]  For each identified table, **write a `SELECT` statement to retrieve the first five records**.
- [ ]  **Note the columns and their respective data types** in each table and understand the information each table contains.
- [ ]  **Consult the embedded data dictionary** within the `data_dictionary` table for explanations of each column.

In [2]:
# loding the SQL extension,connecting to the database and loading the data
from sqlalchemy import create_engine, MetaData, Table, text
import pandas as pd
engine = create_engine("mysql+pymysql://root:Qwerty%4012345@127.0.0.1:3306/md_water_services")
metadata = MetaData()

In [3]:
# Tables in the database
show_tables_query ="""
SHOW TABLES

"""
pd.read_sql_query(show_tables_query, engine)

Tables_in_md_water_services
0               auditor_report
1      combined_analysis_table
2              data_dictionary
3                     employee
4          global_water_access
5            incorrect_records
6                     location
7                       visits
8                water_quality
9                 water_source
10              well_pollution

In [4]:
# Structure of the Tables in the database

queries = {
    "data_dictionary": """
        SELECT *
        FROM data_dictionary
        LIMIT 5;
    """,
    "employee": """
        SELECT *
        FROM employee
        LIMIT 5;
    """,
    "global_water_access": """
        SELECT *
        FROM global_water_access
        LIMIT 10;
    """,
    "location": """
        SELECT *
        FROM location
        LIMIT 10;
    """,
    "visits": """
        SELECT *
        FROM visits
        LIMIT 10;
    """,
    "water_quality": """
        SELECT *
        FROM water_quality
        LIMIT 10;
    """,
    "water_source": """
        SELECT *
        FROM water_source
        LIMIT 10;
    """,
    "well_pollution": """
        SELECT *
        FROM well_pollution
        LIMIT 10;
    """
}

# Loop through tables and display their first 10 rows
for table, query in queries.items():
    print(f"{table.upper()} ")
    display(pd.read_sql_query(query, engine))


DATA_DICTIONARY 


table_name           column_name                          description  \
0   employee  assigned_employee_id  Unique ID assigned to each employee   
1   employee         employee_name                 Name of the employee   
2   employee          phone_number       Contact number of the employee   
3   employee                 email        Email address of the employee   
4   employee               address  Residential address of the employee   

       datatype related_to  
0           INT     visits  
1  VARCHAR(255)             
2   VARCHAR(15)             
3  VARCHAR(255)             
4  VARCHAR(255)

EMPLOYEE 


assigned_employee_id   employee_name  phone_number  \
0                     0     Amara Jengo  +99637993287   
1                     1     Bello Azibo  +99643864786   
2                     2    Bakari Iniko  +99222599041   
3                     3  Malachi Mavuso  +99945849900   
4                     4    Cheche Buhle  +99381679640   

                           email                  address province_name  \
0     amara.jengo@ndogowater.gov  36 Pwani Mchangani Road        Sokoto   
1     bello.azibo@ndogowater.gov    129 Ziwa La Kioo Road      Kilimani   
2    bakari.iniko@ndogowater.gov   18 Mlima Tazama Avenue       Hawassa   
3  malachi.mavuso@ndogowater.gov       100 Mogadishu Road        Akatsi   
4    cheche.buhle@ndogowater.gov         1 Savanna Street        Akatsi   

  town_name        position  
0    Ilanga  Field Surveyor  
1     Rural  Field Surveyor  
2     Rural  Field Surveyor  
3    Lusaka  Field Surveyor  
4     Rural  Field Surveyor

GLOBAL_WATER_ACCESS 


name                      region  year      pop_n   pop_u  \
0     Afghanistan                  South Asia  2015  34413.600  24.803   
1     Afghanistan                  South Asia  2020  38928.300  26.026   
2         Albania       Europe & Central Asia  2015   2890.520  57.434   
3         Albania       Europe & Central Asia  2020   2877.800  62.112   
4         Algeria  Middle East & North Africa  2015  39728.000  70.848   
5         Algeria  Middle East & North Africa  2020  43851.000  73.733   
6  American Samoa         East Asia & Pacific  2015     55.806  87.238   
7  American Samoa         East Asia & Pacific  2020     55.197  87.153   
8         Andorra       Europe & Central Asia  2015     77.993  88.345   
9         Andorra       Europe & Central Asia  2020     77.265  87.916   

   wat_bas_n  wat_lim_n  wat_unimp_n  wat_sur_n  wat_bas_r  wat_lim_r  \
0    61.3398    3.51120    22.168800  12.980200    52.9885    3.86114   
1    75.0914    1.44754    14.560300   8.900780    66.3279    1.95682   
2    93.3943    3.62638     2.979290   0.000000    90.6273    5.26317   
3    95.0680    1.88466     3.047310   0.000000    94.0914    2.30526   
4    93.4096    5.15778     1.275460   0.157193    88.3527    8.68575   
5    94.4373    4.98588     0.531837   0.044952    90.0375    8.79672   
6    99.6191    0.00000     0.380897   0.000000        NaN        NaN   
7    99.7738    0.00000     0.226228   0.000000        NaN        NaN   
8   100.0000    0.00000     0.000002   0.000000   100.0000    0.00000   
9   100.0000    0.00000     0.000000   0.000000   100.0000    0.00000   

   wat_unimp_r  wat_sur_r  wat_bas_u  wat_lim_u  wat_unimp_u  wat_sur_u  
0    26.553300  16.597100    86.6589    2.45027     8.876040   2.014750  
1    19.682900  12.032300   100.0000    0.00000     0.000000   0.000000  
2     4.109550   0.000000    95.4451    2.41331     2.141620   0.000000  
3     3.603380   0.000000    95.6638    1.62809     2.708100   0.000000  
4     2.580430   0.381108    95.4903    3.70612     0.738510   0.065058  
5     0.994603   0.171137    96.0047    3.62829     0.366976   0.000000  
6          NaN        NaN        NaN        NaN          NaN        NaN  
7          NaN        NaN        NaN        NaN          NaN        NaN  
8     0.000000   0.000000   100.0000    0.00000     0.000000   0.000000  
9     0.000000   0.000000   100.0000    0.00000     0.000000   0.000000

LOCATION 


location_id               address province_name town_name location_type
0   AkHa00000    2 Addis Ababa Road        Akatsi    Harare         Urban
1   AkHa00001   10 Addis Ababa Road        Akatsi    Harare         Urban
2   AkHa00002    9 Addis Ababa Road        Akatsi    Harare         Urban
3   AkHa00003  139 Addis Ababa Road        Akatsi    Harare         Urban
4   AkHa00004   17 Addis Ababa Road        Akatsi    Harare         Urban
5   AkHa00005  125 Addis Ababa Road        Akatsi    Harare         Urban
6   AkHa00006   98 Addis Ababa Road        Akatsi    Harare         Urban
7   AkHa00007   21 Addis Ababa Road        Akatsi    Harare         Urban
8   AkHa00008   11 Addis Ababa Road        Akatsi    Harare         Urban
9   AkHa00009    6 Addis Ababa Road        Akatsi    Harare         Urban

VISITS 


record_id location_id     source_id      time_of_record  visit_count  \
0          0   SoIl32582  SoIl32582224 2021-01-01 09:10:00            1   
1          1   KiRu28935  KiRu28935224 2021-01-01 09:17:00            1   
2          2   HaRu19752  HaRu19752224 2021-01-01 09:36:00            1   
3          3   AkLu01628  AkLu01628224 2021-01-01 09:53:00            1   
4          4   AkRu03357  AkRu03357224 2021-01-01 10:11:00            1   
5          5   KiRu29315  KiRu29315224 2021-01-01 10:17:00            1   
6          6   AkRu05234  AkRu05234224 2021-01-01 10:18:00            1   
7          7   KiRu28520  KiRu28520224 2021-01-01 10:28:00            1   
8          8   HaZa21742  HaZa21742224 2021-01-01 10:37:00            1   
9          9   AmDa12214  AmDa12214224 2021-01-01 10:58:00            1   

   time_in_queue  assigned_employee_id  
0             15                    12  
1              0                    46  
2             62                    40  
3              0                     1  
4             28                    14  
5              9                    40  
6              0                    30  
7              0                    34  
8              0                     6  
9              0                    36

WATER_QUALITY 


record_id  subjective_quality_score  visit_count
0          0                         0            1
1          1                         1            1
2          2                         5            1
3          3                        10            1
4          4                         4            1
5          5                         0            1
6          6                         9            1
7          7                        10            1
8          8                         2            1
9          9                        10            1

WATER_SOURCE 


source_id type_of_water_source  number_of_people_served
0  AkHa00000224          tap_in_home                      956
1  AkHa00001224   tap_in_home_broken                      930
2  AkHa00002224   tap_in_home_broken                      486
3  AkHa00003224                 well                      364
4  AkHa00004224   tap_in_home_broken                      942
5  AkHa00005224          tap_in_home                      736
6  AkHa00006224          tap_in_home                      882
7  AkHa00007224          tap_in_home                      554
8  AkHa00008224                 well                      398
9  AkHa00009224                 well                      346

WELL_POLLUTION 


source_id                date  \
0  KiRu28935224 2021-01-04 09:17:00   
1  AkLu01628224 2021-01-04 09:53:00   
2  HaZa21742224 2021-01-04 10:37:00   
3  HaRu19725224 2021-01-04 11:04:00   
4  SoRu35703224 2021-01-04 11:29:00   
5  AkHa00070224 2021-01-04 11:42:00   
6  HaSe21346224 2021-01-04 11:52:00   
7  HaYa21468224 2021-01-04 12:03:00   
8  SoRu36278224 2021-01-04 12:24:00   
9  AkLu02155224 2021-01-04 12:29:00   

                                         description  pollutant_ppm  \
0                          Bacteria: Giardia Lamblia       0.000000   
1                                  Bacteria: E. coli       0.000000   
2  Inorganic contaminants: Zinc, Zinc, Lead, Cadmium       2.715000   
3                                              Clean       0.028859   
4                                  Bacteria: E. coli       0.000000   
5                    Inorganic contaminants: Cadmium       5.467390   
6                                              Clean       0.014038   
7  Inorganic contaminants: Chromium, Barium, Chro...       6.051370   
8                          Parasite: Cryptosporidium       0.000000   
9          Inorganic contaminants: Selenium, Arsenic       7.641060   

   biological                   results  
0  495.898000  Contaminated: Biological  
1    6.096080  Contaminated: Biological  
2    0.000000    Contaminated: Chemical  
3    0.000096                     Clean  
4   22.500900  Contaminated: Biological  
5    0.000000    Contaminated: Chemical  
6    0.000090                     Clean  
7    0.000000    Contaminated: Chemical  
8  485.162000  Contaminated: Biological  
9    0.000000    Contaminated: Chemical


- Database contains **60,000+ unique records** across **43 columns**
- Well-organized tables with intuitive names:
  - `location`: Geographical data (province, town, address)
  - `visits`: Trip logs to water sources
  - `water_source`: Type and capacity of sources
  - Embedded data dictionary explains all columns

### 1.2 Diving into Water Sources

**Context:** This subsection focuses on understanding the different types of water sources recorded in the database and their characteristics. This information is crucial for assessing the scope of the water crisis and planning interventions.

- [ ]  **Identify the table containing information on water source types**.
- [ ]  **Write a SQL query to find all unique types of water sources**.
- [ ]  **Review the descriptions of each water source type**:
    - [ ]  **River**: Open water source with high contamination risk.
    - [ ]  **Well**: Underground source, less contamination risk than rivers, but many are unclean due to aging infrastructure and past corruption.
    - [ ]  **Shared tap**: Public tap shared by communities.
    - [ ]  **Tap in home**: Taps inside citizens' homes, serving about 6 people per tap.
    - [ ]  **Broken tap in home**: In-home taps where infrastructure (pipes, pumps, treatment plants) is non-functional.
- [ ]  **Understand that `tap_in_home` and `tap_in_home_broken` records represent combined data** for multiple households, with `number_of_people_served` reflecting the sum of people in these homes.

In [5]:
water_source_query = """
SELECT 
    *
FROM 
    water_source
LIMIT 10;
"""
pd.read_sql_query(water_source_query, engine)

source_id type_of_water_source  number_of_people_served
0  AkHa00000224          tap_in_home                      956
1  AkHa00001224   tap_in_home_broken                      930
2  AkHa00002224   tap_in_home_broken                      486
3  AkHa00003224                 well                      364
4  AkHa00004224   tap_in_home_broken                      942
5  AkHa00005224          tap_in_home                      736
6  AkHa00006224          tap_in_home                      882
7  AkHa00007224          tap_in_home                      554
8  AkHa00008224                 well                      398
9  AkHa00009224                 well                      346

In [6]:
water_source_query = """
SELECT 
    DISTINCT type_of_water_source
FROM 
    water_source
"""
pd.read_sql_query(water_source_query, engine)

type_of_water_source
0          tap_in_home
1   tap_in_home_broken
2                 well
3           shared_tap
4                river

 **Observation:** 
- There are five unique types of water sources recorded: **tap_in_home, tap_in_home_broken, well, shared_tap, and river**.
-  **Rivers** are open water sources with a high contamination risk, representing the worst possible water source.
- **Wells** draw water from underground, posing less contamination risk than rivers, but many are unclean due to aging infrastructure and past corruption.
-  **Shared taps** are public taps used by communities.
-  **Tap_in_home** refers to taps inside citizens' homes, serving about 6 people per tap on average.
-  **Broken tap_in_home** indicates non-functional in-home taps due to issues with infrastructure like pipes, pumps, or treatment plants.
-  Records for `tap_in_home` and `tap_in_home_broken` combine data for multiple households, with `Number_of_people_served` reflecting the sum of people in these homes (e.g., a single record for `tap_in_home` serving 956 people represents approximately 160 homes).

### 1.3 Unpacking the Visits

**Context:** The `visits` table logs all trips made to different water sources, including visit patterns and queue times. Analyzing this table helps identify locations with significant waiting periods for water access.

- [ ]  **Identify the `visits` table** as the logbook of trips to water sources.
- [ ]  **Retrieve all records where `time_in_queue` is more than 500 minutes** (e.g., 8 hours).
- [ ]  **Note the `source_id` values** from records with long queue times and those with 0 minutes queue time to investigate water source types.
- [ ]  **Check the `water_source` table for the `type_of_water_source`** corresponding to the noted `source_id` values.
- [ ]  **Understand that field surveyors made multiple visits to shared taps** to measure changes in queue times.

In [7]:
# Identifying the Visits table
visits_query = """
SELECT
    *
FROM
    visits
LIMIT 10
"""
pd.read_sql_query(visits_query, engine)

record_id location_id     source_id      time_of_record  visit_count  \
0          0   SoIl32582  SoIl32582224 2021-01-01 09:10:00            1   
1          1   KiRu28935  KiRu28935224 2021-01-01 09:17:00            1   
2          2   HaRu19752  HaRu19752224 2021-01-01 09:36:00            1   
3          3   AkLu01628  AkLu01628224 2021-01-01 09:53:00            1   
4          4   AkRu03357  AkRu03357224 2021-01-01 10:11:00            1   
5          5   KiRu29315  KiRu29315224 2021-01-01 10:17:00            1   
6          6   AkRu05234  AkRu05234224 2021-01-01 10:18:00            1   
7          7   KiRu28520  KiRu28520224 2021-01-01 10:28:00            1   
8          8   HaZa21742  HaZa21742224 2021-01-01 10:37:00            1   
9          9   AmDa12214  AmDa12214224 2021-01-01 10:58:00            1   

   time_in_queue  assigned_employee_id  
0             15                    12  
1              0                    46  
2             62                    40  
3              0                     1  
4             28                    14  
5              9                    40  
6              0                    30  
7              0                    34  
8              0                     6  
9              0                    36

In [8]:
# Records where time_in_queue is more than 8 hours/500 minites

long_wait_query = """
SELECT
    *
FROM
    visits
WHERE
     time_in_queue > 500
"""
pd.read_sql_query(long_wait_query, engine)

record_id location_id     source_id      time_of_record  visit_count  \
0          899   SoRu35083  SoRu35083224 2021-01-16 10:14:00            6   
1         2304   SoKo33124  SoKo33124224 2021-02-06 07:53:00            5   
2         2315   KiRu26095  KiRu26095224 2021-02-06 14:32:00            3   
3         3206   SoRu38776  SoRu38776224 2021-02-20 15:03:00            5   
4         3701   HaRu19601  HaRu19601224 2021-02-27 12:53:00            3   
..         ...         ...           ...                 ...          ...   
100      57408   SoRu35388  SoRu35388224 2023-05-27 08:52:00            5   
101      57832   AkRu04093  AkRu04093224 2023-06-03 07:50:00            3   
102      57843   KiRu30266  KiRu30266224 2023-06-03 16:50:00            2   
103      59129   KiRu27023  KiRu27023224 2023-06-24 16:17:00            2   
104      59597   AmAb09223  AmAb09223224 2023-07-01 13:35:00            2   

     time_in_queue  assigned_employee_id  
0              515                    28  
1              512                    16  
2              529                     8  
3              509                    46  
4              504                     0  
..             ...                   ...  
100            538                     1  
101            524                    34  
102            533                    10  
103            509                     8  
104            509                    24  

[105 rows x 7 columns]

In [9]:
"""
Get water sources with queue times >500 minutes
in here ill join 
Time_in_queue
type_of_water_source on source id
to show the type of water source that has a time in queue greater than 500 minutes
"""
Join_table = """
SELECT
    visits.location_id,
    visits.source_id,
    visits.time_in_queue,
    water_source.type_of_water_source,
    water_source.number_of_people_served

FROM visits
JOIN water_source  
    ON visits.source_id = water_source.source_id
where 
    visits.time_in_queue > 500
ORDER BY
    visits.time_in_queue DESC;
"""
pd.read_sql_query(Join_table, engine)

location_id     source_id  time_in_queue type_of_water_source  \
0     AmRu14612  AmRu14612224            539           shared_tap   
1     HaRu19538  HaRu19538224            539           shared_tap   
2     AkRu05704  AkRu05704224            539           shared_tap   
3     HaRu20126  HaRu20126224            538           shared_tap   
4     SoRu35388  SoRu35388224            538           shared_tap   
..          ...           ...            ...                  ...   
100   AkKi00881  AkKi00881224            502           shared_tap   
101   AmRu14449  AmRu14449224            502           shared_tap   
102   KiRu25672  KiRu25672224            502           shared_tap   
103   AkRu04807  AkRu04807224            501           shared_tap   
104   AmBe11056  AmBe11056224            501           shared_tap   

     number_of_people_served  
0                       3118  
1                       3142  
2                       3398  
3                       3164  
4                       3060  
..                       ...  
100                     3398  
101                     3132  
102                     3334  
103                     3122  
104                     3394  

[105 rows x 5 columns]

In [10]:
"""
in here ill join 
Time_in_queue
type_of_water_source on source id
"""
Join_table = """
SELECT
    visits.source_id,
    visits.time_in_queue,
    water_source.type_of_water_source
FROM visits
JOIN water_source  
    ON visits.source_id = water_source.source_id
where 
    visits.time_in_queue > 500;
"""
pd.read_sql_query(Join_table, engine)

source_id  time_in_queue type_of_water_source
0    SoRu35083224            515           shared_tap
1    SoKo33124224            512           shared_tap
2    KiRu26095224            529           shared_tap
3    SoRu38776224            509           shared_tap
4    HaRu19601224            504           shared_tap
..            ...            ...                  ...
100  SoRu35388224            538           shared_tap
101  AkRu04093224            524           shared_tap
102  KiRu30266224            533           shared_tap
103  KiRu27023224            509           shared_tap
104  AmAb09223224            509           shared_tap

[105 rows x 3 columns]

**Observations** 
- Extreme queue times: **>8 hours** at some shared taps
- Multiple visits logged for shared taps to measure queue changes
- `visit_count > 1` for home taps suggests data errors

### 1.4 Water Source Quality Assessment

**Context:** Assessing the quality of water sources is a critical part of the survey. This involves examining subjective quality scores and identifying potential data errors.

- [ ]  **Identify the `water_quality` table** that contains quality scores for each visit.
- [ ]  **Write a query to find records where `subjective_quality_score` is 10 and `visit_count` indicates a second visit**.
- [ ]  **Identify and note any inconsistencies** or records that should not exist based on survey protocols (e.g., home taps with multiple visits and a perfect score).

In [11]:
visits_query = """
SELECT
    *
FROM
    water_quality
WHERE 
    subjective_quality_score = 10 
    AND visit_count > 1

"""
pd.read_sql_query(visits_query, engine)

record_id  subjective_quality_score  visit_count
0            59                        10            2
1            67                        10            3
2            85                        10            4
3           128                        10            5
4           137                        10            2
...         ...                       ...          ...
1521      60041                        10            6
1522      60053                        10            7
1523      60073                        10            7
1524      60086                        10            8
1525      60094                        10            8

[1526 rows x 3 columns]

In [12]:
#
visits_query = """
SELECT
    wq.subjective_quality_score,
    wq.visit_count,
    ws.type_of_water_source,
    v.record_id
FROM 
    visits v
JOIN 
    water_quality wq ON v.record_id = wq.record_id 
JOIN 
    water_source ws ON v.source_id = ws.source_id
WHERE 
    wq.subjective_quality_score = 10 
    AND wq.visit_count > 1
    

"""
pd.read_sql_query(visits_query, engine)


subjective_quality_score  visit_count type_of_water_source  record_id
0                           10            2           shared_tap         59
1                           10            3           shared_tap         67
2                           10            4           shared_tap         85
3                           10            5           shared_tap        128
4                           10            2           shared_tap        137
...                        ...          ...                  ...        ...
1521                        10            6           shared_tap      60041
1522                        10            7           shared_tap      60053
1523                        10            7           shared_tap      60073
1524                        10            8           shared_tap      60086
1525                        10            8           shared_tap      60094

[1526 rows x 4 columns]


**Observations**
- Quality scores (1-10) show inconsistencies:
  - 218 records with perfect scores (10) but multiple visits
  - Wells marked "Clean" despite biological contamination

### 1.5 Investigating Pollution Issues

**Context:** Pollution status is a major concern, particularly for well sources. This section involves identifying polluted wells, understanding the types of contamination, and correcting data inconsistencies related to pollution records.

- [ ]  **Access the `well_pollution` table** to view contamination data for well sources.
- [ ]  **Understand the classification of wells** as 'Clean', 'Contaminated: Biological', or 'Contaminated: Chemical' based on `results` and `biological` (CFU/mL > 0.01) and `pollutant_ppm` values.
- [ ]  **Write a query to check data integrity** by finding records where `results` is 'Clean' but the `biological` column is `> 0.01`.
- [ ]  **Identify erroneous descriptions** in the `well_pollution` table where "Clean" is mistakenly included in the description alongside biological contamination (e.g., "Clean Bacteria: E. coli", "Clean Bacteria: Giardia Lamblia").
- [ ]  **Prepare SQL `UPDATE` statements to correct these `description` fields**:
    - [ ]  Change `Clean Bacteria: E. coli` to `Bacteria: E. coli`.
    - [ ]  Change `Clean Bacteria: Giardia Lamblia` to `Bacteria: Giardia Lamblia`.
- [ ]  **Prepare SQL `UPDATE` statements to correct `results` column**:
    - [ ]  Change `results` from 'Clean' to 'Contaminated: Biological' where `biological > 0.01` and `results` is 'Clean'.
- [ ]  **Implement updates safely by creating a copy of the `well_pollution` table** (e.g., `well_pollution_copy`) using `CREATE TABLE ... AS (query)` before applying changes.
- [ ]  **Run the `UPDATE` queries on the copy table**.
- [ ]  **Verify that errors are fixed** using a `SELECT` query on the copy table.
- [ ]  If verified, **apply changes to the original `well_pollution` table and drop the copy table**.

In [13]:
well_polution_table = """
SELECT
    *   
FROM
    well_pollution

"""
pd.read_sql_query(well_polution_table, engine)


source_id                date  \
0      KiRu28935224 2021-01-04 09:17:00   
1      AkLu01628224 2021-01-04 09:53:00   
2      HaZa21742224 2021-01-04 10:37:00   
3      HaRu19725224 2021-01-04 11:04:00   
4      SoRu35703224 2021-01-04 11:29:00   
...             ...                 ...   
17378  KiIs23580224 2023-07-10 15:29:00   
17379  AmRu14569224 2023-07-10 15:30:00   
17380  SoMa34164224 2023-07-10 15:43:00   
17381  KiMr24857224 2023-07-10 15:46:00   
17382  AkRu08520224 2023-07-10 16:55:00   

                                             description  pollutant_ppm  \
0                              Bacteria: Giardia Lamblia       0.000000   
1                                      Bacteria: E. coli       0.000000   
2      Inorganic contaminants: Zinc, Zinc, Lead, Cadmium       2.715000   
3                                                  Clean       0.028859   
4                                      Bacteria: E. coli       0.000000   
...                                                  ...            ...   
17378                                              Clean       0.051757   
17379                                              Clean       0.070287   
17380                          Bacteria: Vibrio cholerae       0.000000   
17381              Inorganic contaminants: Lead, Cadmium       3.990230   
17382  Inorganic contaminants: Chromium, Chromium, Ch...       4.386080   

       biological                   results  
0      495.898000  Contaminated: Biological  
1        6.096080  Contaminated: Biological  
2        0.000000    Contaminated: Chemical  
3        0.000096                     Clean  
4       22.500900  Contaminated: Biological  
...           ...                       ...  
17378    0.000084                     Clean  
17379    0.000074                     Clean  
17380  182.880000  Contaminated: Biological  
17381    0.000000    Contaminated: Chemical  
17382    0.000000    Contaminated: Chemical  

[17383 rows x 6 columns]

In [14]:
well_polution_table_query = """
SELECT
    *   
FROM
    well_pollution
WHERE
    results = 'Clean'
    AND biological > 0
"""
pd.read_sql_query(well_polution_table_query, engine)

source_id                date description  pollutant_ppm  biological  \
0     HaRu19725224 2021-01-04 11:04:00       Clean       0.028859    0.000096   
1     HaSe21346224 2021-01-04 11:52:00       Clean       0.014038    0.000090   
2     AkRu05973224 2021-01-04 13:44:00       Clean       0.084344    0.000062   
3     AkHa00706224 2021-01-04 16:10:00       Clean       0.086258    0.000042   
4     HaRu20773224 2021-01-07 09:08:00       Clean       0.059968    0.000024   
...            ...                 ...         ...            ...         ...   
4975  SoIl32597224 2023-07-10 14:53:00       Clean       0.053880    0.000059   
4976  AkRu04063224 2023-07-10 15:04:00       Clean       0.072542    0.000098   
4977  AmAs10214224 2023-07-10 15:08:00       Clean       0.073552    0.000016   
4978  KiIs23580224 2023-07-10 15:29:00       Clean       0.051757    0.000084   
4979  AmRu14569224 2023-07-10 15:30:00       Clean       0.070287    0.000074   

     results  
0      Clean  
1      Clean  
2      Clean  
3      Clean  
4      Clean  
...      ...  
4975   Clean  
4976   Clean  
4977   Clean  
4978   Clean  
4979   Clean  

[4980 rows x 6 columns]

In [15]:
Prep_safe_updt_query = """
SET SQL_SAFE_UPDATES = 0;
"""
with engine.connect() as connection:
    connection.execute(text(Prep_safe_updt_query))

In [16]:
Prep_safe_updt_query = """
CREATE TABLE IF NOT EXISTS
    well_pollution_copy
AS
    SELECT 
        * 
    FROM 
        well_pollution;
"""
with engine.connect() as connection:
    connection.execute(text(Prep_safe_updt_query))

In [17]:
updated_Ecoli_query = """
UPDATE 
    well_pollution_copy
SET
    description = 'Bacteria: E. coli'
WHERE
    description = 'Clean Bacteria: E. coli'
"""
with engine.connect() as connection:
    connection.execute(text(updated_Ecoli_query))


In [18]:
check_query = """
SELECT
    *   
FROM
    well_pollution_copy
WHERE 
    description = 'Bacteria: E. coli'
"""
pd.read_sql_query(check_query, engine)  

source_id                date        description  pollutant_ppm  \
0     AkLu01628224 2021-01-04 09:53:00  Bacteria: E. coli            0.0   
1     SoRu35703224 2021-01-04 11:29:00  Bacteria: E. coli            0.0   
2     KiMr24968224 2021-01-04 13:52:00  Bacteria: E. coli            0.0   
3     AkLu02211224 2021-01-07 12:49:00  Bacteria: E. coli            0.0   
4     HaRu20738224 2021-01-07 14:57:00  Bacteria: E. coli            0.0   
...            ...                 ...                ...            ...   
1255  HaSe21130224 2023-07-08 10:37:00  Bacteria: E. coli            0.0   
1256  HaRu17696224 2023-07-08 12:40:00  Bacteria: E. coli            0.0   
1257  SoRu35478224 2023-07-09 15:42:00  Bacteria: E. coli            0.0   
1258  SoMa33912224 2023-07-10 13:17:00  Bacteria: E. coli            0.0   
1259  KiRu25695224 2023-07-10 14:16:00  Bacteria: E. coli            0.0   

      biological                   results  
0        6.09608  Contaminated: Biological  
1       22.50090  Contaminated: Biological  
2       44.21640  Contaminated: Biological  
3        3.57905  Contaminated: Biological  
4       35.24620  Contaminated: Biological  
...          ...                       ...  
1255    34.23190  Contaminated: Biological  
1256    16.58240  Contaminated: Biological  
1257    30.21620  Contaminated: Biological  
1258     1.33360  Contaminated: Biological  
1259    16.01960  Contaminated: Biological  

[1260 rows x 6 columns]

In [19]:
updated_Giardia_Lamblia_query = """
UPDATE 
    well_pollution_copy
SET
    description = 'Bacteria: Giardia Lamblia'
WHERE
    description = 'Clean Bacteria: Giardia Lamblia'
"""
with engine.connect() as connection:
    connection.execute(text(updated_Giardia_Lamblia_query))

In [20]:
check_query = """
SELECT
    *   
FROM
    well_pollution_copy
WHERE 
    description = 'Bacteria: Giardia Lamblia'
"""
pd.read_sql_query(check_query, engine) 

source_id                date                description  \
0    KiRu28935224 2021-01-04 09:17:00  Bacteria: Giardia Lamblia   
1    KiRu25465224 2021-01-09 10:15:00  Bacteria: Giardia Lamblia   
2    HaRu19788224 2021-01-11 09:16:00  Bacteria: Giardia Lamblia   
3    KiAm22506224 2021-01-11 14:18:00  Bacteria: Giardia Lamblia   
4    SoKo33117224 2021-01-11 16:16:00  Bacteria: Giardia Lamblia   
..            ...                 ...                        ...   
566  HaRu18504224 2023-07-06 15:47:00  Bacteria: Giardia Lamblia   
567  KiRu28978224 2023-07-06 16:15:00  Bacteria: Giardia Lamblia   
568  AmDa12038224 2023-07-07 10:53:00  Bacteria: Giardia Lamblia   
569  SoRu36586224 2023-07-08 12:08:00  Bacteria: Giardia Lamblia   
570  SoRu39412224 2023-07-09 12:29:00  Bacteria: Giardia Lamblia   

     pollutant_ppm  biological                   results  
0              0.0     495.898  Contaminated: Biological  
1              0.0     495.898  Contaminated: Biological  
2              0.0     495.898  Contaminated: Biological  
3              0.0     495.898  Contaminated: Biological  
4              0.0     495.898  Contaminated: Biological  
..             ...         ...                       ...  
566            0.0     495.898  Contaminated: Biological  
567            0.0     495.898  Contaminated: Biological  
568            0.0     495.898  Contaminated: Biological  
569            0.0     495.898  Contaminated: Biological  
570            0.0     495.898  Contaminated: Biological  

[571 rows x 6 columns]

In [21]:
updated_contamination_query = """
UPDATE 
    well_pollution_copy
SET
    results = 'Contaminated: Biological'
WHERE
    biological > 0.01
"""
with engine.connect() as connection:
    connection.execute(text(updated_contamination_query))
    connection.commit() 

In [22]:
check_query = """
SELECT
    *   
FROM
    well_pollution_copy
WHERE 
    biological > 0.01
"""
pd.read_sql_query(check_query, engine) 

source_id                date                description  \
0     KiRu28935224 2021-01-04 09:17:00  Bacteria: Giardia Lamblia   
1     AkLu01628224 2021-01-04 09:53:00          Bacteria: E. coli   
2     SoRu35703224 2021-01-04 11:29:00          Bacteria: E. coli   
3     SoRu36278224 2021-01-04 12:24:00  Parasite: Cryptosporidium   
4     SoRu36313224 2021-01-04 13:46:00  Bacteria: Vibrio cholerae   
...            ...                 ...                        ...   
5369  SoMa33912224 2023-07-10 13:17:00          Bacteria: E. coli   
5370  KiRu25695224 2023-07-10 14:16:00          Bacteria: E. coli   
5371  HaRu19833224 2023-07-10 15:06:00  Parasite: Cryptosporidium   
5372  KiRu28338224 2023-07-10 15:13:00   Virus: Hepatitis A Virus   
5373  SoMa34164224 2023-07-10 15:43:00  Bacteria: Vibrio cholerae   

      pollutant_ppm  biological                   results  
0               0.0   495.89800  Contaminated: Biological  
1               0.0     6.09608  Contaminated: Biological  
2               0.0    22.50090  Contaminated: Biological  
3               0.0   485.16200  Contaminated: Biological  
4               0.0   182.88000  Contaminated: Biological  
...             ...         ...                       ...  
5369            0.0     1.33360  Contaminated: Biological  
5370            0.0    16.01960  Contaminated: Biological  
5371            0.0   485.16200  Contaminated: Biological  
5372            0.0    96.63410  Contaminated: Biological  
5373            0.0   182.88000  Contaminated: Biological  

[5374 rows x 6 columns]

In [23]:
drop_query = """
DROP TABLE IF EXISTS
    well_pollution
"""
with engine.connect() as connection:
    connection.execute(text(drop_query))
    connection.commit() 

In [24]:
Rename_query = """
ALTER TABLE 
    well_pollution_copy
RENAME TO 
    well_pollution;
"""
with engine.connect() as connection:
    connection.execute(text(Rename_query))
    connection.commit() 

In [25]:
# Tables in the database
show_tables_query ="""
SHOW TABLES

"""
pd.read_sql_query(show_tables_query, engine)

Tables_in_md_water_services
0               auditor_report
1      combined_analysis_table
2              data_dictionary
3                     employee
4          global_water_access
5            incorrect_records
6                     location
7                       visits
8                water_quality
9                 water_source
10              well_pollution

- **Observation:** The `well_pollution` table provides contamination data specifically for well sources, including `date`, `description`, `pollutant_ppm`, `biological`, and `results`.
- **Observation:** Wells are classified as 'Clean', 'Contaminated: Biological', or 'Contaminated: Chemical' based on `results` and `biological` (CFU/mL > 0.01) and `pollutant_ppm` values.
- **Observation:** There is **data integrity issue**: some records show `results` as 'Clean' even when the `biological` column is > 0.01 (indicating contamination). This suggests data input personnel misinterpreted the `description` field.
- **Observation:** Erroneous `description` fields exist, such as "Clean Bacteria: E. coli" or "Clean Bacteria: Giardia Lamblia", where "Clean" is mistakenly included despite biological contamination.
- **Observation:** These inconsistencies require **correction** by updating the `description` fields (e.g., to "Bacteria: E. coli") and changing `results` from 'Clean' to 'Contaminated: Biological' when `biological` > 0.01. Changes are recommended to be tested on a copy table first for safety.

# Data Cleaning and Performance Analysis

This section focuses on further refining the collected data, including updating employee information and analysing employee performance in the survey. It also initiates a deeper analysis of the locations of water sources and the characteristics of water sources themselves, setting the stage for problem identification.

### 2.1 Cleaning Employee Data

**Context:** Ensuring the accuracy and completeness of employee contact information is essential for communication and project management. This involves standardising email addresses and phone numbers.

- [ ]  **Update the `email` column in the `employee` table** by constructing email addresses in the format `first_name.last_name@ndogowater.gov` using `REPLACE()`, `LOWER()`, and `CONCAT()` functions.
- [ ]  **Identify and remove any trailing spaces from phone numbers** in the `phone_number` column using the `TRIM()` function to ensure proper length and format.

In [26]:
employee_Table_query = """
SELECT 
    *
FROM
    employee
LIMIT 10;
"""
pd.read_sql_query(employee_Table_query, engine)

assigned_employee_id   employee_name  phone_number  \
0                     0     Amara Jengo  +99637993287   
1                     1     Bello Azibo  +99643864786   
2                     2    Bakari Iniko  +99222599041   
3                     3  Malachi Mavuso  +99945849900   
4                     4    Cheche Buhle  +99381679640   
5                     5  Zuriel Matembo  +99034075111   
6                     6    Deka Osumare  +99379364631   
7                     7  Lalitha Kaburi  +99681623240   
8                     8     Enitan Zuri  +99248509202   
9                    10       Farai Nia  +99570082739   

                           email                    address province_name  \
0     amara.jengo@ndogowater.gov    36 Pwani Mchangani Road        Sokoto   
1     bello.azibo@ndogowater.gov      129 Ziwa La Kioo Road      Kilimani   
2    bakari.iniko@ndogowater.gov     18 Mlima Tazama Avenue       Hawassa   
3  malachi.mavuso@ndogowater.gov         100 Mogadishu Road        Akatsi   
4    cheche.buhle@ndogowater.gov           1 Savanna Street        Akatsi   
5  zuriel.matembo@ndogowater.gov   26 Bahari Ya Faraja Road      Kilimani   
6    deka.osumare@ndogowater.gov        104 Kenyatta Street        Akatsi   
7  lalitha.kaburi@ndogowater.gov  145 Sungura Amanpour Road      Kilimani   
8     enitan.zuri@ndogowater.gov           117 Kampala Road       Hawassa   
9       farai.nia@ndogowater.gov  33 Angélique Kidjo Avenue        Amanzi   

  town_name        position  
0    Ilanga  Field Surveyor  
1     Rural  Field Surveyor  
2     Rural  Field Surveyor  
3    Lusaka  Field Surveyor  
4     Rural  Field Surveyor  
5     Rural  Field Surveyor  
6     Rural  Field Surveyor  
7     Rural  Field Surveyor  
8  Zanzibar  Field Surveyor  
9    Dahabu  Field Surveyor

In [27]:
updt_email = """
UPDATE 
    employee
SET 
    email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov');
"""
with engine.connect() as connection:
    connection.execute(text(query))
    connection.commit() 

In [28]:
employee_Table_query2 = """
SELECT 
    *
FROM
    employee
LIMIT 10;
"""
pd.read_sql_query(employee_Table_query2, engine)


assigned_employee_id   employee_name  phone_number  \
0                     0     Amara Jengo  +99637993287   
1                     1     Bello Azibo  +99643864786   
2                     2    Bakari Iniko  +99222599041   
3                     3  Malachi Mavuso  +99945849900   
4                     4    Cheche Buhle  +99381679640   
5                     5  Zuriel Matembo  +99034075111   
6                     6    Deka Osumare  +99379364631   
7                     7  Lalitha Kaburi  +99681623240   
8                     8     Enitan Zuri  +99248509202   
9                    10       Farai Nia  +99570082739   

                           email                    address province_name  \
0     amara.jengo@ndogowater.gov    36 Pwani Mchangani Road        Sokoto   
1     bello.azibo@ndogowater.gov      129 Ziwa La Kioo Road      Kilimani   
2    bakari.iniko@ndogowater.gov     18 Mlima Tazama Avenue       Hawassa   
3  malachi.mavuso@ndogowater.gov         100 Mogadishu Road        Akatsi   
4    cheche.buhle@ndogowater.gov           1 Savanna Street        Akatsi   
5  zuriel.matembo@ndogowater.gov   26 Bahari Ya Faraja Road      Kilimani   
6    deka.osumare@ndogowater.gov        104 Kenyatta Street        Akatsi   
7  lalitha.kaburi@ndogowater.gov  145 Sungura Amanpour Road      Kilimani   
8     enitan.zuri@ndogowater.gov           117 Kampala Road       Hawassa   
9       farai.nia@ndogowater.gov  33 Angélique Kidjo Avenue        Amanzi   

  town_name        position  
0    Ilanga  Field Surveyor  
1     Rural  Field Surveyor  
2     Rural  Field Surveyor  
3    Lusaka  Field Surveyor  
4     Rural  Field Surveyor  
5     Rural  Field Surveyor  
6     Rural  Field Surveyor  
7     Rural  Field Surveyor  
8  Zanzibar  Field Surveyor  
9    Dahabu  Field Surveyor

In [29]:
Prep_safe_updt_query = """
UPDATE 
    employee
SET 
    phone_number = TRIM(phone_number)
"""
with engine.connect() as connection:
    connection.execute(text(Prep_safe_updt_query))
    connection.commit() 

In [30]:
trim_check = """
SELECT 
    phone_number,
    length(phone_number) AS phone_length
FROM
    employee

"""
pd.read_sql_query(trim_check, engine)

phone_number  phone_length
0   +99637993287            12
1   +99643864786            12
2   +99222599041            12
3   +99945849900            12
4   +99381679640            12
5   +99034075111            12
6   +99379364631            12
7   +99681623240            12
8   +99248509202            12
9   +99570082739            12
10  +99438283153            12
11  +99917230124            12
12  +99313603182            12
13  +99820596465            12
14  +99176320477            12
15  +99611183730            12
16  +99690308755            12
17  +99522694997            12
18  +99797494944            12
19  +99822478933            12
20  +99533195846            12
21  +99046972648            12
22  +99477692836            12
23  +99895654233            12
24  +99613194771            12
25  +99489752164            12
26  +99269319001            12
27  +99239172440            12
28  +99239164858            12
29  +99578401238            12
30  +99345162378            12
31  +99239271483            12
32  +99738251476            12
33  +99471523698            12
34  +99381254763            12
35  +99142537684            12
36  +99924513876            12
37  +99568247135            12
38  +99735681472            12
39  +99317854629            12
40  +99874512369            12
41  +99127436598            12
42  +99583124796            12
43  +99273841596            12
44  +99968217435            12
45  +99682314579            12
46  +99721534698            12
47  +99578123469            12
48  +99375418692            12
49  +99178325649            12
50  +99234156789            12
51  +99712584936            12
52  +99634157892            12
53  +99231569847            12
54  +99123456978            12
55  +99412356789            12

- **Observation:** Employee email addresses in the `employee` table are missing and need to be generated in a standard format: `first_name.last_name@ndogowater.gov`.
- **Observation:** `phone_number` entries in the `employee` table have **trailing spaces**, causing their length to be 13 characters instead of the expected 12. These need to be trimmed.

### 2.2 Honouring the Workers

**Context:** Recognizing the effort of field workers is important. This involves assessing their geographical distribution and identifying top performers based on the number of locations visited during the survey.

- [ ]  **Count the number of employees living in each town** from the `employee` table, grouping by `town_name`.
- [ ]  **Identify the top 3 field surveyors** with the most location visits by counting `records` in the `visits` table grouped by `assigned_employee_id` and ordering in descending order.
- [ ]  **Retrieve the names, email addresses, and phone numbers** of these top 3 surveyors from the `employee` table using their `assigned_employee_id`.

In [31]:
employee_table = """
SELECT
    *
FROM
    employee


"""
pd.read_sql_query(employee_table, engine)

assigned_employee_id     employee_name  phone_number  \
0                      0       Amara Jengo  +99637993287   
1                      1       Bello Azibo  +99643864786   
2                      2      Bakari Iniko  +99222599041   
3                      3    Malachi Mavuso  +99945849900   
4                      4      Cheche Buhle  +99381679640   
5                      5    Zuriel Matembo  +99034075111   
6                      6      Deka Osumare  +99379364631   
7                      7    Lalitha Kaburi  +99681623240   
8                      8       Enitan Zuri  +99248509202   
9                     10         Farai Nia  +99570082739   
10                    12       Gamba Shani  +99438283153   
11                    14      Harith Nyota  +99917230124   
12                    16       Isoke Amani  +99313603182   
13                    18     Jengo Tumaini  +99820596465   
14                    20        Kunto Asha  +99176320477   
15                    22       Lesedi Kofi  +99611183730   
16                    24      Makena Thabo  +99690308755   
17                    26        Nia Furaha  +99522694997   
18                    28          Ona Sefu  +99797494944   
19                    30         Pili Zola  +99822478933   
20                    32     Quibilah Akua  +99533195846   
21                    34        Rudo Imani  +99046972648   
22                    36     Sanaa Tendaji  +99477692836   
23                    38     Thandiwe Kito  +99895654233   
24                    40         Usafi Ayo  +99613194771   
25                    42    Vuyisile Kwame  +99489752164   
26                    44     Wambui Jabali  +99269319001   
27                    46        Xola Uzuri  +99239172440   
28                    48     Yewande Ebele  +99239164858   
29                    49      Nadia Zalika  +99578401238   
30                    50          Ayo Deka  +99345162378   
31                    52     Bahira Jamila  +99239271483   
32                    53       Chidi Kunto  +99738251476   
33                    54     Dalila Lesedi  +99471523698   
34                    56       Eshe Makena  +99381254763   
35                    58   Faridah Khalida  +99142537684   
36                    62        Gavivi Ona  +99924513876   
37                    64       Hasani Pili  +99568247135   
38                    68    Iniko Quibilah  +99735681472   
39                    71        Jengo Rudo  +99317854629   
40                    73        Kofi Sanaa  +99874512369   
41                    75     Lulu Thandiwe  +99127436598   
42                    79      Makena Usafi  +99583124796   
43                    81      Nia Vuyisile  +99273841596   
44                    83        Ona Wambui  +99968217435   
45                    87         Pili Xola  +99682314579   
46                    89  Quibilah Yewande  +99721534698   
47                    93       Rudo Zalika  +99578123469   
48                    94         Sanaa Ayo  +99375418692   
49                    96       Thabo Akida  +99178325649   
50                    98       Usafi Bahir  +99234156789   
51                   102   Vuyisile Ghadir  +99712584936   
52                   104     Wambui Hadiya  +99634157892   
53                   105        Xola Inaya  +99231569847   
54                   108     Yewande Jengo  +99123456978   
55                   110         Zuri Kofi  +99412356789   

                              email                     address province_name  \
0        amara.jengo@ndogowater.gov     36 Pwani Mchangani Road        Sokoto   
1        bello.azibo@ndogowater.gov       129 Ziwa La Kioo Road      Kilimani   
2       bakari.iniko@ndogowater.gov      18 Mlima Tazama Avenue       Hawassa   
3     malachi.mavuso@ndogowater.gov          100 Mogadishu Road        Akatsi   
4       cheche.buhle@ndogowater.gov            1 Savanna Street        Akatsi   
5     zuriel.matembo@ndogowater.gov    26 Bahari Ya Faraja Road      Kilimani   
6       deka.osu

In [32]:
employee_count = """
SELECT
    COUNT(*) AS employee_count,
    town_name
FROM
    employee
GROUP BY
    town_name

"""
pd.read_sql_query(employee_count, engine)

employee_count town_name
0               3    Ilanga
1              29     Rural
2               4    Lusaka
3               4  Zanzibar
4               6    Dahabu
5               1  Kintampo
6               5    Harare
7               1   Yaounde
8               3    Serowe

In [33]:
employee_table = """
SELECT
    *
FROM
    VISITS
lIMIT 5;


"""
pd.read_sql_query(employee_table, engine)

record_id location_id     source_id      time_of_record  visit_count  \
0          0   SoIl32582  SoIl32582224 2021-01-01 09:10:00            1   
1          1   KiRu28935  KiRu28935224 2021-01-01 09:17:00            1   
2          2   HaRu19752  HaRu19752224 2021-01-01 09:36:00            1   
3          3   AkLu01628  AkLu01628224 2021-01-01 09:53:00            1   
4          4   AkRu03357  AkRu03357224 2021-01-01 10:11:00            1   

   time_in_queue  assigned_employee_id  
0             15                    12  
1              0                    46  
2             62                    40  
3              0                     1  
4             28                    14

In [34]:
top_surveyor = """
SELECT 
    e.employee_name,
    e.phone_number,
    e.email,
    e.assigned_employee_id,
    SUM(v.visit_count) AS total_visits
FROM
    employee e
JOIN
    visits v ON e.assigned_employee_id = v.assigned_employee_id
GROUP BY
    e.assigned_employee_id, e.employee_name, e.phone_number, e.email
ORDER BY
    total_visits DESC
;
"""
pd.read_sql_query(top_surveyor, engine)

employee_name  phone_number                          email  \
0      Bello Azibo  +99643864786     bello.azibo@ndogowater.gov   
1        Pili Zola  +99822478933       pili.zola@ndogowater.gov   
2       Rudo Imani  +99046972648      rudo.imani@ndogowater.gov   
3   Malachi Mavuso  +99945849900  malachi.mavuso@ndogowater.gov   
4        Farai Nia  +99570082739       farai.nia@ndogowater.gov   
5      Enitan Zuri  +99248509202     enitan.zuri@ndogowater.gov   
6   Zuriel Matembo  +99034075111  zuriel.matembo@ndogowater.gov   
7    Sanaa Tendaji  +99477692836   sanaa.tendaji@ndogowater.gov   
8    Yewande Ebele  +99239164858   yewande.ebele@ndogowater.gov   
9         Ona Sefu  +99797494944        ona.sefu@ndogowater.gov   
10     Gamba Shani  +99438283153     gamba.shani@ndogowater.gov   
11  Vuyisile Kwame  +99489752164  vuyisile.kwame@ndogowater.gov   
12       Usafi Ayo  +99613194771       usafi.ayo@ndogowater.gov   
13   Thandiwe Kito  +99895654233   thandiwe.kito@ndogowater.gov   
14    Makena Thabo  +99690308755    makena.thabo@ndogowater.gov   
15    Bakari Iniko  +99222599041    bakari.iniko@ndogowater.gov   
16   Quibilah Akua  +99533195846   quibilah.akua@ndogowater.gov   
17   Jengo Tumaini  +99820596465   jengo.tumaini@ndogowater.gov   
18      Xola Uzuri  +99239172440      xola.uzuri@ndogowater.gov   
19    Harith Nyota  +99917230124    harith.nyota@ndogowater.gov   
20  Lalitha Kaburi  +99681623240  lalitha.kaburi@ndogowater.gov   
21    Deka Osumare  +99379364631    deka.osumare@ndogowater.gov   
22     Amara Jengo  +99637993287     amara.jengo@ndogowater.gov   
23    Cheche Buhle  +99381679640    cheche.buhle@ndogowater.gov   
24      Nia Furaha  +99522694997      nia.furaha@ndogowater.gov   
25     Isoke Amani  +99313603182     isoke.amani@ndogowater.gov   
26   Wambui Jabali  +99269319001   wambui.jabali@ndogowater.gov   
27     Lesedi Kofi  +99611183730     lesedi.kofi@ndogowater.gov   
28      Kunto Asha  +99176320477      kunto.asha@ndogowater.gov   

    assigned_employee_id  total_visits  
0                      1        8944.0  
1                     30        8800.0  
2                     34        8411.0  
3                      3        8152.0  
4                     10        8111.0  
5                      8        7747.0  
6                      5        7708.0  
7                     36        7393.0  
8                     48        6573.0  
9                     28        6542.0  
10                    12        6229.0  
11                    42        5688.0  
12                    40        5592.0  
13                    38        5061.0  
14                    24        4871.0  
15                     2        4805.0  
16                    32        4484.0  
17                    18        4129.0  
18                    46        3704.0  
19                    14        3513.0  
20                     7        3051.0  
21                     6        2806.0  
22                     0        2779.0  
23                     4        2648.0  
24                    26        1783.0  
25                    16        1754.0  
26                    44         470.0  
27                    22         339.0  
28                    20          43.0

- **Observation:** A significant number of **field workers reside in smaller, rural communities** across Maji Ndogo.
- **Observation:** The **top 3 field surveyors** with the most location visits can be identified from the `visits` table, allowing for recognition of their efforts.

### 2.3 Analysing Locations

**Context:** Understanding the geographical distribution of water sources across Maji Ndogo's provinces and towns is crucial for strategic planning. This involves counting records by location type and calculating percentages to gain clearer insights.

- [ ]  **Count the number of records per town** using the `location` table.
- [ ]  **Count the number of records per province** using the `location` table.
- [ ]  **Create a result set showing `province_name`, `town_name`, and `records_per_town`**, grouped by both province and town, and ordered primarily by `province_name` and then by `records_per_town` in descending order.
- [ ]  **Count the number of records for each `location_type`** (Urban/Rural).
- [ ]  **Calculate the percentage of water sources in rural communities** from the total sources.
- [ ]  **Note key insights from the location analysis**:
    - [ ]  The entire country was properly surveyed, ensuring the dataset represents the situation on the ground.
    - [ ]  60% of water sources are located in rural communities, which is a key consideration for future decisions.

In [35]:
location_Table = """
SELECT
    *
FROM
    location    
"""
pd.read_sql_query(location_Table, engine)

location_id               address province_name town_name location_type
0       AkHa00000    2 Addis Ababa Road        Akatsi    Harare         Urban
1       AkHa00001   10 Addis Ababa Road        Akatsi    Harare         Urban
2       AkHa00002    9 Addis Ababa Road        Akatsi    Harare         Urban
3       AkHa00003  139 Addis Ababa Road        Akatsi    Harare         Urban
4       AkHa00004   17 Addis Ababa Road        Akatsi    Harare         Urban
...           ...                   ...           ...       ...           ...
39645   SoRu39645  183 Angelique Street        Sokoto     Rural         Rural
39646   SoRu39646   89 Angelique Street        Sokoto     Rural         Rural
39647   SoRu39647  116 Angelique Street        Sokoto     Rural         Rural
39648   SoRu39648    9 Angelique Street        Sokoto     Rural         Rural
39649   SoRu39649  150 Angelique Street        Sokoto     Rural         Rural

[39650 rows x 5 columns]

In [36]:
location_records_town = """
SELECT
    town_name, 
    COUNT(*) AS records_per_town 
FROM 
    location 
GROUP BY 
    town_name 
ORDER BY 
    records_per_town DESC
"""
pd.read_sql_query(location_records_town, engine)

town_name  records_per_town
0      Rural             23740
1     Harare              1650
2      Amina              1090
3     Lusaka              1070
4     Mrembo               990
5     Asmara               930
6     Dahabu               930
7   Kintampo               780
8     Ilanga               780
9    Isiqalo               770
10      Zuri               750
11     Amara               710
12   Majengo               620
13     Pwani               520
14      Kofi               510
15    Serowe               480
16    Bahari               470
17    Marang               460
18   Abidjan               400
19     Bello               400
20    Cheche               370
21      Deka               350
22    Djenne               310
23  Zanzibar               300
24   Yaounde               270

In [37]:
location_records_province = """
SELECT
    province_name, 
    COUNT(*) AS records_per_town 
FROM 
    location 
GROUP BY 
    province_name
ORDER BY 
    records_per_town DESC
"""
pd.read_sql_query(location_records_province, engine)

province_name  records_per_town
0      Kilimani              9510
1        Akatsi              8940
2        Sokoto              8220
3        Amanzi              6950
4       Hawassa              6030

In [38]:
location_type = """
SELECT 
    location_type, 
    COUNT(*) AS num_sources
FROM location
GROUP BY location_type;
"""
pd.read_sql_query(location_type, engine)

location_type  num_sources
0         Urban        15910
1         Rural        23740

In [39]:
# Detailed records by province and town
location_type = """

SELECT 
    province_name,
    town_name,
    COUNT(*) AS records_per_town
FROM 
    location
GROUP BY 
    province_name, town_name
ORDER BY 
    province_name, records_per_town DESC;
"""
pd.read_sql_query(location_type, engine)

province_name town_name  records_per_town
0         Akatsi     Rural              6290
1         Akatsi    Lusaka              1070
2         Akatsi    Harare               800
3         Akatsi  Kintampo               780
4         Amanzi     Rural              3100
5         Amanzi    Asmara               930
6         Amanzi    Dahabu               930
7         Amanzi     Amina               670
8         Amanzi     Pwani               520
9         Amanzi   Abidjan               400
10        Amanzi     Bello               400
11       Hawassa     Rural              3900
12       Hawassa    Serowe               480
13       Hawassa     Amina               420
14       Hawassa      Deka               350
15       Hawassa    Djenne               310
16       Hawassa  Zanzibar               300
17       Hawassa   Yaounde               270
18      Kilimani     Rural              5440
19      Kilimani    Mrembo               990
20      Kilimani    Harare               850
21      Kilimani   Isiqalo               770
22      Kilimani      Zuri               750
23      Kilimani     Amara               710
24        Sokoto     Rural              5010
25        Sokoto    Ilanga               780
26        Sokoto   Majengo               620
27        Sokoto      Kofi               510
28        Sokoto    Bahari               470
29        Sokoto    Marang               460
30        Sokoto    Cheche               370

In [40]:
# Percentage of rural water sources
location_type = """
WITH location_counts AS (
    SELECT 
        location_type,
        COUNT(*) AS count
    FROM 
        location
    GROUP BY 
        location_type
)
SELECT 
    ROUND(
        (SELECT count FROM location_counts WHERE location_type = 'Rural') / 
        SUM(count) * 100, 
        2
    ) AS rural_percentage
FROM 
    location_counts;
"""
pd.read_sql_query(location_type, engine)

rural_percentage
0             59.87

- **Observation:** The survey thoroughly covered the entire country, with a **similar number of water sources documented in each province**, ensuring the dataset represents the situation on the ground.
- **Observation:** **60% of water sources are located in rural communities** across Maji Ndogo, which is a key consideration for future planning and interventions.

### 2.4 Diving into the Sources

**Context:** This subsection aims to understand the scope of the water crisis by analyzing the `water_source` table. This involves calculating total people surveyed, counting source types, determining average people served per source, and calculating the total and percentage of people served by each source type.

- [ ]  **Calculate the total number of people surveyed** across all water sources.
- [ ]  **Count how many of each different `type_of_water_source` there are** and sort the results.
- [ ]  **Calculate the average number of people served by each `type_of_water_source`**, remembering to interpret the results for `tap_in_home` and `tap_in_home_broken` by dividing by 6 (average people per household).
- [ ]  **Calculate the total number of people served by each `type_of_water_source`** and order them with the most people served at the top.
- [ ]  **Convert the total people served by each source type into percentages**, rounding to 0 decimals, using the overall total population surveyed.
- [ ]  **Identify key insights from the source analysis**:
    - [ ]  43% of the population uses shared taps, with an average of 2000 people sharing one.
    - [ ]  31% of people have in-home water infrastructure, but 45% of these are non-functional.
    - [ ]  18% of people use wells, but only 28% of these wells are clean.

In [1]:
people_served_query = """
SELECT 
    SUM(number_of_people_served) AS total_population_surveyed
FROM 
    water_source;
"""
pd.read_sql_query(people_served_query, engine)

NameError: name 'pd' is not defined

In [ ]:
count_water_source = """
SELECT 
    type_of_water_source,
    COUNT(*) AS number_of_sources
FROM 
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    number_of_sources DESC
"""
pd.read_sql_query(count_water_source, engine)

type_of_water_source  number_of_sources
0                 well              17383
1          tap_in_home               7265
2   tap_in_home_broken               5856
3           shared_tap               5767
4                river               3379

In [ ]:
avg_ppl_served = """
SELECT 
    type_of_water_source,
    CASE 
        WHEN type_of_water_source IN ('tap_in_home', 'tap_in_home_broken') 
        THEN ROUND(AVG(number_of_people_served)/6, 0)
        ELSE ROUND(AVG(number_of_people_served), 0)
    END AS avg_people_served
FROM 
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    avg_people_served DESC;
"""
pd.read_sql_query(avg_ppl_served, engine)

type_of_water_source  avg_people_served
0           shared_tap             2071.0
1                river              699.0
2                 well              279.0
3   tap_in_home_broken              108.0
4          tap_in_home              107.0

In [ ]:
pop_served_per_source = """
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served
FROM 
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    total_people_served DESC;
"""
pd.read_sql_query(pop_served_per_source, engine)

type_of_water_source  total_people_served
0           shared_tap           11945272.0
1                 well            4841724.0
2          tap_in_home            4678880.0
3   tap_in_home_broken            3799720.0
4                river            2362544.0

In [ ]:
perc_pop_served = """
WITH total_population AS (
    SELECT SUM(number_of_people_served) AS total
    FROM water_source
),
source_counts AS (
    SELECT 
        type_of_water_source,
        SUM(number_of_people_served) AS people_served
    FROM 
        water_source
    GROUP BY 
        type_of_water_source
)
SELECT 
    sc.type_of_water_source,
    sc.people_served,
    ROUND((sc.people_served / tp.total) * 100, 0) AS percentage_of_population
FROM 
    source_counts sc,
    total_population tp
ORDER BY 
    sc.people_served DESC;
"""
pd.read_sql_query(perc_pop_served, engine)

type_of_water_source  people_served  percentage_of_population
0           shared_tap     11945272.0                      43.0
1                 well      4841724.0                      18.0
2          tap_in_home      4678880.0                      17.0
3   tap_in_home_broken      3799720.0                      14.0
4                river      2362544.0                       9.0

- **Observation:** The total number of people surveyed across all water sources is approximately **27 million citizens**.
- **Observation:** Shared taps and in-home taps are the most numerous source types.
- **Observation:** On average, **644 people share a `tap_in_home`** and 649 people share a `tap_in_home_broken` record, which actually represents many households combined, with an average of 6 people per household. Conversely, **2071 people share a single public `shared_tap`** on average, explaining the long queues.
- **Observation:** **43% of the population uses shared taps**, making them the most common source of water.
- **Observation:** **31% of people have in-home water infrastructure**, but **45% of these are non-functional** due to issues with pipes, pumps, or treatment plants. Towns like Amina and rural Amanzi have significant broken infrastructure.
- **Observation:** **18% of people use wells**, but **only 28% of these wells are clean** (from previous analysis). Most of these contaminated wells are in Hawassa, Kilimani, and Akatsi.

# Strategic Planning and Problem Solving

This section transitions from data analysis to formulating actionable plans and identifying areas for intervention. It involves ranking water sources by impact, understanding queue time patterns, and addressing issues of data integrity and potential corruption within the survey teams.

### 3.1 Start of a Solution: Ranking Sources

**Context:** To prioritize interventions, water sources are ranked based on the number of people they serve, focusing on improvable sources. This helps in making data-driven decisions on where to allocate resources for repairs and upgrades.

- [ ]  **Rank each `type_of_water_source` based on the total number of people served**, excluding `tap_in_home` as it represents the best available source.
- [ ]  **Create a query to assign a priority rank to individual water sources within each type**, based on `number_of_people_served`, considering only improvable sources.
- [ ]  **Experiment with different ranking functions (RANK(), DENSE_RANK(), ROW_NUMBER())** to understand their implications for prioritisation lists.
- [ ]  **Consider what information engineers would need** beyond `source_id` to locate and repair sources (e.g., addresses).

In [46]:
ranking_water_sources = """
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) AS priority_rank
FROM 
    water_source
WHERE 
    type_of_water_source != 'tap_in_home'
GROUP BY 
    type_of_water_source
ORDER BY 
    priority_rank;
"""
pd.read_sql_query(ranking_water_sources, engine)

type_of_water_source  total_people_served  priority_rank
0           shared_tap           11945272.0              1
1                 well            4841724.0              2
2   tap_in_home_broken            3799720.0              3
3                river            2362544.0              4

In [ ]:
prioritizing_water_sources = """
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) AS priority_rank
FROM 
    water_source
WHERE 
    type_of_water_source != 'tap_in_home'
GROUP BY 
    type_of_water_source
ORDER BY 
    priority_rank;
"""
pd.read_sql_query(prioritizing_water_sources , engine)

- **Observation:** To prioritise interventions, water sources should be ranked based on the total number of people they serve, excluding `tap_in_home` as it represents the best available source not needing improvement.
- **Observation:** **Shared taps** should be prioritised for improvement, followed by wells and broken in-home taps, as they affect the most people.
- **Observation:** Different ranking functions (`RANK()`, `DENSE_RANK()`, `ROW_NUMBER()`) have different implications for prioritization lists (e.g., `ROW_NUMBER()` provides unique ranks even for sources with equal `number_of_people_served`).
- **Observation:** Engineers would need **address information** beyond just `source_id` to locate and repair sources effectively.

### 3.2 Analysing Queues

**Context:** Understanding queue patterns helps to identify peak times and days for water collection, which can inform short-term solutions like deploying water tankers. This analysis uses DateTime functions to extract daily and hourly patterns.

- [ ]  **Calculate the total duration of the survey** by finding the difference between the minimum and maximum `time_of_record` in the `visits` table.
- [ ]  **Calculate the average total queue time** in Maji Ndogo, excluding records where `time_in_queue` is 0 (e.g., taps in homes) using `NULLIF()`.
- [ ]  **Calculate the average queue time for each day of the week** using `DAYNAME(time_of_record)`, grouped by day.
- [ ]  **Calculate the average queue time for each hour of the day**, formatting the hour to `'%H:00'` for better readability.
- [ ]  **Create a pivot table** showing average queue times for each hour across all days of the week, using `CASE()` statements for each day column.
- [ ]  **Identify patterns in queue times**:
    - [ ]  Long queues on Monday mornings and evenings.
    - [ ]  Lowest queue times on Wednesdays, but long queues on Wednesday evenings.
    - [ ]  Queue times on Saturdays are approximately twice as long as weekdays.
    - [ ]  Shortest queues are on Sundays due to cultural priorities.

In [ ]:
survey_duration = """
SELECT 
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS total_survey_days,
    CONCAT(
        FLOOR(DATEDIFF(MAX(time_of_record), MIN(time_of_record)) / 365), ' years ',
        FLOOR((DATEDIFF(MAX(time_of_record), MIN(time_of_record)) %% 365) / 30), ' months'
    ) AS survey_duration
FROM 
    visits;
"""
pd.read_sql_query(survey_duration, engine)

In [ ]:
queue_duration = """
SELECT 
    ROUND(AVG(NULLIF(time_in_queue, 0)), 1) AS avg_queue_time_minutes,
    CONCAT(
        FLOOR(AVG(NULLIF(time_in_queue, 0)) / 60), ' hours ',
        ROUND(MOD(AVG(NULLIF(time_in_queue, 0)), 60)), ' minutes'
    ) AS avg_queue_time_formatted
FROM 
    visits
WHERE 
    time_in_queue > 0;
"""
pd.read_sql_query(queue_duration, engine)

In [ ]:
avg_weekly_queue_duration = """
SELECT 
    DAYNAME(time_of_record) AS day_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0)), 0) AS avg_queue_time_minutes,
    CONCAT(
        FLOOR(AVG(NULLIF(time_in_queue, 0)) / 60), 'h ',
        ROUND(MOD(AVG(NULLIF(time_in_queue, 0)), 60)), 'm'
    ) AS avg_queue_time_formatted
FROM 
    visits
WHERE 
    time_in_queue > 0
GROUP BY 
    day_of_week
ORDER BY 
    FIELD(day_of_week, 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday');
"""
pd.read_sql_query(avg_weekly_queue_duration, engine)

In [ ]:
avg_hourly_queue_duration = """
SELECT 
    TIME_FORMAT(TIME(time_of_record), '%%H:00') AS hour_of_day,
    ROUND(AVG(NULLIF(time_in_queue, 0)), 1) AS avg_queue_time_minutes
FROM 
    visits
WHERE 
    time_in_queue > 0
GROUP BY 
    hour_of_day
ORDER BY 
    hour_of_day;
"""
pd.read_sql_query(avg_hourly_queue_duration, engine)

In [ ]:
avg_queue_times_table = """
SELECT 
    TIME_FORMAT(TIME(time_of_record), '%%H:00') AS hour_of_day,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Sunday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Monday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Tuesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Wednesday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Thursday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Friday,
    ROUND(AVG(CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN NULLIF(time_in_queue, 0) ELSE NULL END), 1) AS Saturday
FROM 
    visits
WHERE 
    time_in_queue > 0
GROUP BY 
    hour_of_day
ORDER BY 
    hour_of_day;
"""
pd.read_sql_query(avg_queue_times_table, engine)

- **Observation:** The water survey took approximately **924 days (about 2.5 years)** to complete.
- **Observation:** The **average total queue time** for water in Maji Ndogo is about **123 minutes (over 2 hours)**, excluding sources with 0 queue times (like functional home taps).
- **Observation:** **Saturday** has significantly longer average queue times (approx. twice as long as weekdays).
- **Observation:** Queue times are **longest on Monday mornings and evenings** and **Wednesday evenings**, reflecting peak collection times.
- **Observation:** **Wednesdays and Sundays have the shortest queue times** overall, with Sundays being particularly short due to cultural priorities (family and religion).
- **Observation:** Formatting hour numbers as '%H:00' (e.g., '06:00') improves readability for time-based analysis.
- **Observation:** Pivot tables (constructed using `CASE()` statements) are useful for showing average queue times for each hour across all days of the week, revealing detailed patterns.

### 3.3 Integrating Auditor's Report and Uncovering Corruption

**Context:** The independent auditor's report highlighted inconsistencies in survey data. This crucial step involves integrating the audit findings, comparing surveyor and auditor scores, and linking incorrect records to specific employees to investigate potential misconduct.

- [ ]  **Load the `auditor_report.csv` file** into a new SQL table named `auditor_report`.
- [ ]  **Join `auditor_report` with `visits` and `water_quality` tables** to compare `true_water_source_score` from the auditor with `subjective_quality_score` from surveyors, renaming columns for clarity and filtering for `visits.visit_count = 1`.
- [ ]  **Identify records where `surveyor_score` does not match `auditor_score`** (the 102 incorrect records).
- [ ]  **Verify that `type_of_water_source` data remains consistent** between the audit report and survey data despite score discrepancies.
- [ ]  **Join the `employee` table to the identified incorrect records** to link errors to specific `employee_name`.
- [ ]  **Create a `VIEW` named `Incorrect_records`** containing `location_id`, `record_id`, `employee_name`, `auditor_score`, `surveyor_score`, and `statements` for all mismatched records.
- [ ]  **Create a CTE named `error_count`** to calculate the number of mistakes made by each employee from the `Incorrect_records` view.
- [ ]  **Calculate the average number of mistakes per employee** from the `error_count` CTE.
- [ ]  **Create a CTE named `suspect_list`** to identify employees who made an `above-average number of mistakes`.
- [ ]  **Filter the `Incorrect_records` view** to display only records associated with employees in the `suspect_list`, including their `statements`.
- [ ]  **Filter `statements` for keywords like "cash"** to find incriminating remarks related to corruption.
- [ ]  **Cross-check if any employee outside the `suspect_list` has statements mentioning "cash"**.
- [ ]  **Summarize the evidence for suspected corrupt employees** (higher error rates and incriminating statements).

In [ ]:
modify_auditor_report = """
ALTER TABLE 
    auditor_report
MODIFY COLUMN location_id VARCHAR(20),
MODIFY COLUMN type_of_water_source VARCHAR(50),
MODIFY COLUMN true_water_source_score INT,
MODIFY COLUMN statements TEXT;
"""
with engine.connect() as connection:
    connection.execute(text(modify_auditor_report))
    connection.commit()

In [ ]:
check_auditor_report = """
SELECT 
    COUNT(DISTINCT v.record_id) AS unique_visits,
    COUNT(DISTINCT wq.record_id) AS unique_water_quality_records,
    COUNT(*) AS total_records
FROM 
    visits v
JOIN 
    water_quality wq ON v.record_id = wq.record_id;
"""
pd.read_sql_query(check_auditor_report, engine)

In [ ]:
compare_surv_audit = """
-- Create view for incorrect records
CREATE OR REPLACE VIEW Incorrect_records AS
SELECT 
    ar.location_id,
    v.record_id,
    e.employee_name,
    ar.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    ar.statements
FROM 
    auditor_report ar
JOIN 
    visits v ON ar.location_id = v.location_id
JOIN 
    water_quality wq ON v.record_id = wq.record_id
JOIN 
    employee e ON v.assigned_employee_id = e.assigned_employee_id
WHERE 
    v.visit_count = 1
    AND ar.true_water_source_score != wq.subjective_quality_score;
"""
with engine.connect() as connection:
    connection.execute(text(compare_surv_audit))
    connection.commit()

In [ ]:
check_compare_surv_audit = """
SELECT
    *
FROM
     Incorrect_records
"""
pd.read_sql_query(check_compare_surv_audit, engine)

In [ ]:
verfy_water_source = """
SELECT 
    COUNT(*) AS inconsistent_source_types
FROM 
    auditor_report ar
JOIN 
    water_source ws ON ar.location_id = ws.source_id
WHERE 
    ar.type_of_water_source != ws.type_of_water_source;
"""
pd.read_sql_query(verfy_water_source, engine)

In [ ]:
 Identified_Suspect = """
-- Calculate error counts per employee
WITH error_count AS (
    SELECT 
        employee_name,
        COUNT(*) AS number_of_errors
    FROM 
        Incorrect_records
    GROUP BY 
        employee_name
),
avg_errors AS (
    SELECT 
        AVG(number_of_errors) AS avg_error_count
    FROM 
        error_count
)
SELECT 
    ec.employee_name,
    ec.number_of_errors,
    ae.avg_error_count
FROM 
    error_count ec,
    avg_errors ae
WHERE 
    ec.number_of_errors > ae.avg_error_count
ORDER BY 
    ec.number_of_errors DESC;
"""
pd.read_sql_query( Identified_Suspect, engine)

In [ ]:
evidence = """
-- Create suspect list
WITH suspect_list AS (
    SELECT employee_name
    FROM Incorrect_records
    GROUP BY employee_name
    HAVING COUNT(*) > (SELECT AVG(error_count) 
                      FROM (SELECT COUNT(*) AS error_count 
                            FROM Incorrect_records 
                            GROUP BY employee_name) AS avg_errors)
)
SELECT 
    ir.employee_name,
    ir.location_id,
    ir.auditor_score,
    ir.surveyor_score,
    ir.statements
FROM 
    Incorrect_records ir
JOIN 
    suspect_list sl ON ir.employee_name = sl.employee_name
WHERE 
    ir.statements LIKE '%%cash%%'
    OR ir.statements LIKE '%%bribe%%'
    OR ir.statements LIKE '%%money%%'
ORDER BY 
    ir.employee_name;
"""
pd.read_sql_query(evidence, engine)

In [ ]:
corruption_report = """

WITH suspect_errors AS (
    SELECT 
        employee_name,
        COUNT(*) AS error_count,
        GROUP_CONCAT(DISTINCT location_id) AS affected_locations,
        GROUP_CONCAT(DISTINCT statements SEPARATOR ' | ') AS incriminating_statements
    FROM 
        Incorrect_records
    WHERE 
        employee_name IN ('Zuriel Matembo', 'Malachi Mavuso', 'Bello Azibo', 'Lalitha Kaburi')
        AND (statements LIKE '%%cash%%' OR statements LIKE '%%bribe%%' OR statements LIKE '%%money%%')
    GROUP BY 
        employee_name
)
SELECT 
    se.employee_name,
    e.phone_number,
    e.email,
    e.province_name,
    se.error_count,
    ROUND(se.error_count * 100.0 / (SELECT COUNT(*) FROM Incorrect_records), 1) AS percentage_of_errors,
    se.affected_locations,
    se.incriminating_statements
FROM 
    suspect_errors se
JOIN 
    employee e ON se.employee_name = e.employee_name;
"""

pd.read_sql_query(corruption_report, engine)

- **Observation:** An independent auditor's report was integrated into the database to verify data integrity and accuracy.
- **Observation:** The `visits` table serves as the central link between `location`, `water_quality`, and `employee` tables. The relationship between `visits` and `water_quality` was initially shown as many-to-one but was corrected to **one-to-one**.
- **Observation:** The auditor checked 1620 sites, and **94% (1518 out of 1620) of the records were correct**, indicating high data alignment with governance principles.
- **Observation:** **102 records were found to be incorrect** when comparing the `true_water_source_score` from the auditor to the `subjective_quality_score` from surveyors, filtering for `visits.visit_count = 1`.
- **Observation:** Despite score discrepancies, the **`type_of_water_source` data remained consistent** between the audit report and survey data, validating previous analyses that relied on source types.
- **Observation:** Incorrect records could be **linked to specific employees** by joining `employee` and `visits` tables, revealing who recorded the erroneous data.
- **Observation:** A **`VIEW` named `Incorrect_records`** was created to simplify future queries involving mismatched auditor and surveyor scores, including employee names and auditor statements.
- **Observation:** Analysis showed that **some surveyors made a significantly higher number of "mistakes"** than others.
- **Observation:** A **`CTE` named `suspect_list`** was created to identify employees who made an **above-average number of mistakes**.
- **Observation:** Filtering `Incorrect_records` by employees in the `suspect_list` and searching their `statements` for keywords like "cash" revealed **incriminating remarks related to corruption**, and these allegations were only present against the identified suspects.
- **Observation:** The evidence, while not decisive proof, is concerning enough to be flagged, as President Naledi has a strong stance against corruption.

# Final Analysis and Practical Implementation

This final section focuses on bringing together all insights to formulate a comprehensive plan for improving water access in Maji Ndogo. It involves creating a combined analysis table, performing detailed geographical analysis of water source issues, and designing a `Project_progress` table for tracking future interventions.

### 4.1 Joining Pieces Together: Combined Analysis Table

**Context:** To facilitate comprehensive analysis, data from various tables (`location`, `water_source`, `visits`, `well_pollution`) needs to be assembled into a single, cohesive view. This `combined_analysis_table` will serve as the foundation for identifying specific problems in towns and provinces.

- [ ]  **Identify necessary columns** from `location` (`town_name`, `province_name`, `location_type`), `water_source` (`type_of_water_source`, `number_of_people_served`), `visits` (`time_in_queue`, `visit_count`), and `well_pollution` (`results`).
- [ ]  **Join `location` to `visits`** on `location_id`.
- [ ]  **Join `water_source` to `visits`** on `source_id`.
- [ ]  **Filter the results to include only records where `visits.visit_count = 1`** to avoid duplicate source information from multiple visits.
- [ ]  **Use a `LEFT JOIN` to include `well_pollution.results`** data, as it only applies to wells and should show `NULL` for other source types.
- [ ]  **Create a `VIEW` named `combined_analysis_table`** with the assembled data for simplified future queries.

In [ ]:
combined_analysis_table = """
CREATE OR REPLACE VIEW combined_analysis_table AS
SELECT 
    -- Location information
    l.location_id,
    l.province_name,
    l.town_name,
    l.location_type,
    
    -- Water source details
    ws.source_id,
    ws.type_of_water_source,
    ws.number_of_people_served,
    
    -- Visit information
    v.time_in_queue,
    v.visit_count,
    v.record_id,
    
    -- Well pollution data (only for wells)
    wp.results AS well_pollution_results,
    
    -- Calculated fields
    CASE 
        WHEN ws.type_of_water_source IN ('tap_in_home', 'tap_in_home_broken') 
        THEN ROUND(ws.number_of_people_served / 6, 0) -- Estimate actual taps
        ELSE ws.number_of_people_served
    END AS adjusted_people_served,
    
    CASE 
        WHEN wp.results = 'Contaminated: Biological' THEN 'Biological'
        WHEN wp.results = 'Contaminated: Chemical' THEN 'Chemical'
        WHEN wp.results = 'Clean' THEN 'Clean'
        ELSE NULL
    END AS contamination_type
FROM 
    visits v
JOIN 
    location l ON v.location_id = l.location_id
JOIN 
    water_source ws ON v.source_id = ws.source_id
LEFT JOIN 
    well_pollution wp ON ws.source_id = wp.source_id
WHERE 
    v.visit_count = 1; -- Only include first visit to avoid duplicates
"""
with engine.connect() as connection:
    connection.execute(text(combined_analysis_table))
    connection.commit()

In [ ]:
check_combined_analysis = """
SELECT
    *
FROM
    combined_analysis_table
"""
pd.read_sql_query(check_combined_analysis, engine)

In [ ]:
query = """
-- Check distribution of water source types
SELECT 
    type_of_water_source,
    COUNT(*) AS count,
    ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM combined_analysis_table), 1) AS percentage
FROM 
    combined_analysis_table
GROUP BY 
    type_of_water_source
ORDER BY 
    count DESC;

"""
pd.read_sql_query(query, engine)

In [ ]:
query = """
SELECT 
    type_of_water_source,
    contamination_type,
    COUNT(*) AS count
FROM 
    combined_analysis_table
WHERE 
    type_of_water_source = 'well'
GROUP BY 
    type_of_water_source, contamination_type;
"""
pd.read_sql_query(query, engine)

- **Observation:** A **`VIEW` named `combined_analysis_table`** was created by joining `location`, `water_source`, `visits`, and `well_pollution` tables.
- **Observation:** This view includes `town_name`, `province_name`, `location_type` (from `location`), `type_of_water_source`, `number_of_people_served` (from `water_source`), `time_in_queue` (from `visits`), and `well_pollution.results` (from `well_pollution`).
- **Observation:** Filtering for `visits.visit_count = 1` is crucial to avoid duplicate source information from multiple visits to the same location, ensuring accurate aggregations.
- **Observation:** A `LEFT JOIN` was used for `well_pollution.results` because pollution data only applies to wells, resulting in `NULL` values for other source types, which is the desired outcome.
- **Observation:** This combined table serves as the foundation for identifying specific problems in towns and provinces.



### 4.2 The Last Analysis: Regional Problem Identification

**Context:** Using the `combined_analysis_table`, pivot tables are created to break down water source issues by province and town. This detailed geographical analysis helps pinpoint specific areas where certain water problems (e.g., river water use, broken taps, shared tap queues) are most prevalent.

- [ ]  **Create a CTE `province_totals`** to calculate the total `people_served` for each `province_name`.
- [ ]  **Generate a pivot table aggregating data per province** (using `province_totals` and `combined_analysis_table`), showing the percentage of people served by each `source_type` (`river`, `shared_tap`, `tap_in_home`, `tap_in_home_broken`, `well`) per province.
- [ ]  **Identify provincial patterns** such as:
    - [ ]  Sokoto having the largest population drinking river water, indicating a priority for drilling wells.
    - [ ]  Amanzi having a majority of taps, but half being broken, indicating a need for infrastructure repair.
- [ ]  **Create a CTE `town_totals`** to calculate the total `people_served` for each `province_name` and `town_name` (necessary due to duplicate town names across provinces).
- [ ]  **Generate a pivot table aggregating data per town** (using `town_totals` and `combined_analysis_table`), showing the percentage of people served by each `source_type` per town.
- [ ]  **Store the town-aggregated data in a `TEMPORARY TABLE`** called `town_aggregated_water_access` for faster access and further calculations.
- [ ]  **Analyze the town-level data** to identify specific local issues:
    - [ ]  High disparities in water access in Sokoto (some drinking river water while others have home taps), suggesting unequal wealth distribution.
    - [ ]  Amina (in Amanzi) having a high percentage of broken in-home taps, making it a priority for infrastructure repair.
    - [ ]  Calculate the percentage of `broken_taps` relative to all `tap_in_home` infrastructure to highlight towns with poor functionality.

In [ ]:
provincial_analysis = """
-- Create province-level pivot table with percentages
WITH province_totals AS (
    SELECT 
        province_name,
        SUM(number_of_people_served) AS total_ppl_serv
    FROM 
        combined_analysis_table
    GROUP BY 
        province_name
)
SELECT 
    cat.province_name,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'river' THEN cat.number_of_people_served ELSE 0 END) / pt.total_ppl_serv * 100, 0) AS river_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'shared_tap' THEN cat.number_of_people_served ELSE 0 END) / pt.total_ppl_serv * 100, 0) AS shared_tap_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'tap_in_home' THEN cat.number_of_people_served ELSE 0 END) / pt.total_ppl_serv * 100, 0) AS tap_in_home_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'tap_in_home_broken' THEN cat.number_of_people_served ELSE 0 END) / pt.total_ppl_serv * 100, 0) AS tap_in_home_broken_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'well' THEN cat.number_of_people_served ELSE 0 END) / pt.total_ppl_serv * 100, 0) AS well_percentage,
    pt.total_ppl_serv
FROM 
    combined_analysis_table cat
JOIN 
    province_totals pt ON cat.province_name = pt.province_name
GROUP BY 
    cat.province_name, pt.total_ppl_serv
ORDER BY 
    cat.province_name;
"""
pd.read_sql_query(provincial_analysis, engine)

In [ ]:
Town_analysis = """
-- Then create the temporary table
CREATE TEMPORARY TABLE town_aggregated_water_access AS
WITH town_totals AS (
    SELECT 
        province_name,
        town_name,
        SUM(number_of_people_served) AS total_ppl_serv
    FROM 
        combined_analysis_table
    GROUP BY 
        province_name, town_name
)
SELECT 
    cat.province_name,
    cat.town_name,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'river' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS river_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'shared_tap' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS shared_tap_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'tap_in_home' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS tap_in_home_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'tap_in_home_broken' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS tap_in_home_broken_percentage,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'well' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS well_percentage,
    tt.total_ppl_serv,
    ROUND(AVG(CASE WHEN cat.type_of_water_source IN ('shared_tap', 'well', 'river') THEN cat.time_in_queue ELSE NULL END), 1) AS avg_queue_time,
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'well' AND cat.contamination_type IN ('Biological', 'Chemical') THEN cat.number_of_people_served ELSE 0 END) / 
          NULLIF(SUM(CASE WHEN cat.type_of_water_source = 'well' THEN cat.number_of_people_served ELSE 0 END), 0) * 100, 0) AS well_contamination_percentage
FROM 
    combined_analysis_table cat
JOIN 
    town_totals tt ON cat.province_name = tt.province_name AND cat.town_name = tt.town_name
GROUP BY 
    cat.province_name, cat.town_name, tt.total_ppl_serv
ORDER BY 
    cat.province_name, cat.town_name;
"""

with engine.connect() as connection:
    connection.execute(text(Town_analysis))
    connection.commit()



In [ ]:
query = """
-- Query the temporary table for key insights
SELECT 
    * 
FROM 
    town_aggregated_water_access
ORDER BY 
    river_percentage DESC, tap_in_home_broken_percentage DESC
LIMIT 20;
"""
pd.read_sql_query(query, engine)

In [ ]:
query = """
-- Calculate broken tap percentages by town
SELECT 
    province_name,
    town_name,
    tap_in_home_percentage,
    tap_in_home_broken_percentage,
    ROUND(tap_in_home_broken_percentage / NULLIF(tap_in_home_percentage + tap_in_home_broken_percentage, 0) * 100, 0) AS broken_tap_percentage,
    total_ppl_serv
FROM 
    town_aggregated_water_access
WHERE 
    (tap_in_home_percentage + tap_in_home_broken_percentage) > 0
ORDER BY 
    broken_tap_percentage DESC
LIMIT 10;
"""
pd.read_sql_query(query, engine)

In [ ]:
query = """
CREATE TEMPORARY TABLE town_aggregated_water_access AS
WITH town_totals AS (
    SELECT 
        province_name,
        town_name,
        SUM(number_of_people_served) AS total_ppl_serv
    FROM 
        combined_analysis_table
    GROUP BY 
        province_name, town_name
)
SELECT 
    cat.province_name,
    cat.town_name,
    -- Existing columns (unchanged)...
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'well' THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS well_percentage,
    
    -- NEW: Percentage served by CONTAMINATED wells only
    ROUND(SUM(CASE WHEN cat.type_of_water_source = 'well' AND cat.contamination_type IN ('Biological', 'Chemical') 
              THEN cat.number_of_people_served ELSE 0 END) / tt.total_ppl_serv * 100, 0) AS dirty_well_percentage,
    
    -- Rest of the query remains the same...
FROM 
    combined_analysis_table cat
JOIN 
    town_totals tt ON cat.province_name = tt.province_name AND cat.town_name = tt.town_name
GROUP BY 
    cat.province_name, cat.town_name, tt.total_ppl_serv;
"""
pd.read_sql_query(query, engine)

**The Last Analysis: Regional Problem Identification**

- **Observation:** **Sokoto province** has the **largest population drinking river water**, indicating it should be a priority for drilling new wells (specifically rural parts and Bahari city).
- **Observation:** In **Amanzi province, a majority of water comes from taps, but half of these are broken**, indicating a need for infrastructure repair.
- **Observation:** **Amina town (in Amanzi)** has a high percentage of broken in-home taps (over 50% of people have taps installed but not working), making it a top priority for infrastructure repair. Fixing these will also reduce queue times.
- **Observation:** High disparities in water access, such as some people drinking river water while others have home taps in Sokoto, suggest **unequal wealth distribution**.
- **Observation:** Duplicate town names (e.g., Harare, Amina) across different provinces necessitate **grouping by both `province_name` and `town_name`** for accurate town-level aggregation.
- **Observation:** Storing aggregated town-level data in a **`TEMPORARY TABLE` (e.g., `town_aggregated_water_access`)** improves query performance for further calculations.
- **Observation:** Other towns with significant shared tap issues (long queues) include Bello, Abidjan, and Zuri. Towns with broken infrastructure issues also include Lusaka, Zuri, Djenne, and rural Amanzi.

**4.3 Practical Plan: Database Implementation**

- **Observation:** A `Project_progress` table was created to implement and track the water crisis intervention plan.
- **Observation:** The `Project_progress` table includes columns for `Project_id` (SERIAL PRIMARY KEY), `source_id` (VARCHAR, NOT NULL, REFERENCES `water_source`), `Address`, `Town`, `Province`, `Source_type`, `Improvement`, `Source_status` (VARCHAR, DEFAULT 'Backlog', CHECK ('Backlog', 'In progress', 'Complete')), `Date_of_completion`, and `Comments`.
- **Observation:** The plan prioritises improving sources that affect the most people, focusing on shared taps, contaminated wells, and existing broken infrastructure, while `tap_in_home` (working) sources are not a current focus.
- **Observation:** Specific improvement actions are defined based on source type and conditions:
    - **Rivers**: 'Drill wells'.
    - **Wells**: 'Install RO filter' for chemical contamination and 'Install UV and RO filter' for biological contamination.
    - **Shared taps** (queue time >= 30 minutes): `CONCAT('Install ', FLOOR(time_in_queue / 30), ' taps nearby')` (e.g., install one tap for every 30 minutes of queue time, rounded down).
    - **Broken tap in home**: 'Diagnose local infrastructure'.
- **Observation:** The populated `Project_progress` table will provide clear, data-driven directions for engineers and allow for tracking of improvements.